what4/src/What4/Serialize/Normalize.hs

```diff
@@ -0,0 +1,158 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE MultiWayIf #-}
+
+-- | Normalization and equivalence checking for expressions
+module What4.Serialize.Normalize
+  ( normSymFn
+  , normExpr
+  , testEquivSymFn
+  , testEquivExpr
+  , ExprEquivResult(..)
+  ) where
+
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.TraversableFC as FC
+
+import qualified What4.Interface as S
+import qualified What4.Expr as S
+import qualified What4.Expr.Builder as B
+import qualified What4.Expr.WeightedSum as WSum
+import           Data.Parameterized.Classes
+
+
+normSymFn :: forall sym st fs t args ret. sym ~ B.ExprBuilder t st fs
+          => sym
+          -> B.ExprSymFn t args ret
+          -> Ctx.Assignment (S.Expr t) args
+          -> IO (S.Expr t ret)
+normSymFn sym symFn argEs = case B.symFnInfo symFn of
+  B.DefinedFnInfo argBVs expr _ -> do
+    argEs' <- FC.traverseFC (normExpr sym) argEs
+    expr' <- B.evalBoundVars sym expr argBVs argEs'
+    normExpr sym expr'
+  _ -> S.applySymFn sym symFn argEs
+
+
+normExpr :: forall sym st fs t tp
+          . sym ~ B.ExprBuilder t st fs
+         => sym
+         -> B.Expr t tp -> IO (B.Expr t tp)
+normExpr sym e = go e
+  where go :: B.Expr t tp -> IO (B.Expr t tp)
+        go (B.SemiRingLiteral S.SemiRingIntegerRepr val _) = S.intLit sym val
+        go (B.AppExpr appExpr) = normAppExpr sym appExpr
+        go x@(B.NonceAppExpr nae) =
+          case B.nonceExprApp nae of
+            B.FnApp fn args -> normSymFn sym fn args
+            _ -> return x
+        go x = return x
+
+-- | Normalize an expression by passing it back through the builder
+-- FIXME: incomplete
```

Comment:
Worth filing an issue about this? I'm not about the sense in which this in incomplete.

Reply:
I'm also not sure what is incomplete here, if anything